File: impl_java/modelGeneration/folModel/rules/OrElim.java
package modelGeneration.folModel.rules;

import fol.formula.And;
import fol.formula.Formula;
import fol.formula.Not;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import modelGeneration.folModel.Model;

public class OrElim implements InferenceRule {

    @Override
    public boolean apply(Model model) {
        var or = getOr(model);
        if(or.isEmpty()) return false;
        var maybeApplication = getRuleApplication(or.get(), model);
        if (maybeApplication.isEmpty()) return false;
        var application = maybeApplication.get();
        model.addExtensions(Map.of(application, application.outputExtensions()));
        model.addRuleApplications(Set.of(application));
        return true;
    }

    @Override
    public boolean isBranching() {
        return true;
    }

    private Optional<Not> getOr(Model model) {
        return model.getFormulas().stream()
                .filter(f -> f instanceof Not && ((Not) f).formula() instanceof And)
                .map(Not.class::cast).findAny();
    }

    private Optional<OrElimRuleApplication> getRuleApplication(Not or, Model model) {
        Set<RuleApplication> modelRuleApplications = model.getRuleApplications();
        // Empty rule application marks that the formula was already expanded
        OrElimRuleApplication emptyApplication = new OrElimRuleApplication(or, Set.of());
        if (modelRuleApplications.contains(emptyApplication)) {
            return Optional.empty();
        }
        Formula left = ((And) or.formula()).left();
        // Add negation and simplify double negation
        left = left instanceof Not(Formula newLeft) ? newLeft : new Not(left);
        Formula right = ((And) or.formula()).right();
        right = right instanceof Not(Formula newRight) ? newRight : new Not(right);
        // Create left and right extensions
        Model leftExtension = new Model(model.getFormulas(), model.getSentenceDepth(), model.getIndividuals(), Map.of(), model.getRuleApplications(), model.getParent());
        leftExtension.addFormulas(Set.of(left));
        leftExtension.setParent(model);
        Model rightExtension = new Model(model.getFormulas(), model.getSentenceDepth(), model.getIndividuals(), Map.of(), model.getRuleApplications(), model.getParent());
        rightExtension.addFormulas(Set.of(right));
        rightExtension.setParent(model);
        OrElimRuleApplication application = new OrElimRuleApplication(or, Set.of(leftExtension, rightExtension));
        leftExtension.addRuleApplications(Set.of(emptyApplication));
        rightExtension.addRuleApplications(Set.of(emptyApplication));
        return Optional.of(application);
    }

    public record OrElimRuleApplication(Not input, Set<Model> outputExtensions) implements RuleApplication {
        @Override
        public String toString() {
            return getString(0, "*");

        }

        @Override
        public boolean equals(Object obj) {
            if (obj instanceof OrElimRuleApplication(Not otherInput, Set<Model> _)) {
                return this.input.equals(otherInput);
            }
            return false;
        }

        @Override
        public String getString(int indentation, String delim) {
            if(indentation == 0) {
                StringBuilder sb = new StringBuilder("OrElim : ");
                sb.append(input).append(" -> \n");
                for (Model model : outputExtensions) {
                    sb.append(model.createString(1, delim)).append("\n");
                }
                return sb.toString();
            }
            String outerIndent = "  ".repeat(Math.max(0, indentation)) + delim + " ";
            StringBuilder sb = new StringBuilder(outerIndent)
                    .append("OrElim : ")
                    .append(input);
            if (!outputExtensions.isEmpty()) {
                sb.append(" -> \n");
            }
            for (Model model : outputExtensions) {
                sb.append(model.createString(indentation + 1, delim)).append("\n");
            }
            return sb.toString();
        }

        @Override
        public int hashCode() {
            return input.hashCode();
        }
    }
}

File: python_port/src/nls_python/model_generation/inference_rules/or_elim.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Tuple

from nls_python.typing_utils import typechecked

from nls_python.fol.formula.fand import And
from nls_python.fol.formula.fnot import Not

from ..tableau_model import TableauModel
from .base import InferenceRule, RuleApplication


@typechecked
class OrElim(InferenceRule):
    @property
    def branching(self) -> bool:
        return True

    def apply(self, model: TableauModel) -> bool:
        candidate = self._find_disjunction(model)
        if candidate is None:
            return False
        application = self._build_application(candidate, model)
        if application is None:
            return False
        model.add_extensions({application: set(application.output_extensions)})
        model.add_rule_applications({application})
        return True

    def _find_disjunction(self, model: TableauModel) -> Optional[Not]:
        for formula in model.get_formulas():
            if isinstance(formula, Not) and isinstance(formula.inner, And):
                return formula
        return None

    def _build_application(
        self, disjunction: Not, model: TableauModel
    ) -> Optional["OrElimApplication"]:
        empty_application = OrElimApplication(disjunction, tuple())
        if empty_application in model.get_rule_applications():
            return None

        left = disjunction.inner.left
        left = left.inner if isinstance(left, Not) else Not(left)
        right = disjunction.inner.right
        right = right.inner if isinstance(right, Not) else Not(right)

        left_extension = model.branch_copy()
        left_extension.add_formulas({left})
        left_extension.set_parent(model)
        left_extension.add_rule_applications({empty_application})

        right_extension = model.branch_copy()
        right_extension.add_formulas({right})
        right_extension.set_parent(model)
        right_extension.add_rule_applications({empty_application})

        return OrElimApplication(disjunction, (left_extension, right_extension))


@typechecked
@dataclass(frozen=True)
class OrElimApplication(RuleApplication):
    source: Not
    output_extensions: Tuple[TableauModel, ...]

    def __hash__(self) -> int:
        return hash(self.source)

    def __eq__(self, other: object) -> bool:
        return isinstance(other, OrElimApplication) and self.source == other.source

    def describe(self, indentation: int = 0, delim: str = "*") -> str:
        indent = "  " * max(indentation, 0) + delim + " "
        lines = [f"{indent}OrElim: {self.source} ->"]
        for idx, branch in enumerate(self.output_extensions, start=1):
            branch_header = "  " * (indentation + 1) + delim + f" Branch {idx}:"
            lines.append(branch_header)
            lines.append(branch.create_string(indentation + 2, delim))
        return "\n".join(lines)

    def __str__(self) -> str:
        return f"OrElim: {self.source}"
